Dshop/apps/products_catalogue/tests/test_products_variants.py

```diff
@@ -0,0 +1,108 @@
+        price=11,
+        short_description="short desc",
+        full_description="full_description"
+    )
+
+
+@pytest.mark.django_db
+def test_product_creation_with_variants():
+    category = Category.objects.create(name='Test Category', is_active=True)
+    main_product = Product.objects.create(
+        name='main_product',
+        category=category,
+        price=10,
+        short_description='short desc',
+        full_description="full desc",
+    )
+    Product.objects.create(
+        name="child product 1",
+        category=category,
+        price=11,
+        short_description="short desc",
+        full_description="full_description",
+        parent_product=main_product
+    )
+    Product.objects.create(
+        name="child product 2",
+        category=category,
+        price=11,
+        short_description="short desc",
+        full_description="full_description",
+        parent_product=main_product
+    )
+    assert Product.objects.count() == 3
+
+
+@pytest.mark.django_db
+def test_lowest_price_in_30_days_for_product_variant():
+    with freeze_time("2023-07-01"):
+        category = Category.objects.create(name='Test Category', is_active=True)
+        product = Product.objects.create(
+            name='main_product',
+            category=category,
+            price=10,
+            short_description='short desc',
+            full_description="full desc",
+        )
+        variant = Product.objects.create(
+            name="child product",
+            category=category,
+            price=11,
+            short_description="short desc",
+            full_description="full_description",
+            parent_product=product
+        )
+    with freeze_time("2023-07-28"):
+        variant.price = 0.05
+        variant.save()
+    with freeze_time("2023-11-02"):
+        variant.price = 0.25
+        variant.save()
+    with freeze_time("2023-11-03"):
+        variant.price = 1
+        variant.save()
+    with freeze_time("2023-11-04"):
+        assert variant.lowest_price_in_30_days == pytest.approx(Decimal('0.05'))
+
+
+@pytest.mark.django_db
+def test_product_variant_attributes():
+    category = Category.objects.create(name='Test Category', is_active=True)
+    main_product = Product.objects.create(
+        name='main_product',
+        category=category,
+        price=10,
+        short_description='short desc',
+        full_description="full desc",
+    )
+    variant1 = Product.objects.create(
+        name="child product 1",
+        category=category,
+        price=11,
+        short_description="short desc",
+        full_description="full_description",
+        parent_product=main_product
+    )
+    ProductAttribute.objects.create(
+        product=variant1,
+        key='color',
+        value='red'
+    )
+    ProductAttribute.objects.create(
+        product=variant1,
+        key='size',
+        value='XL'
+    )
+    assert variant1.get_attributes().count() == 2
```

Comment:
Would be good to add some more tests
* I don't see any test here checking the relation between main product and variants. Example: create product, add variants to it, and then show that {% for variant in product_variants %} in template makes sense. You can test view and verify data in context. Examples: https://github.com/pymasterspl/Dshop/blob/a359913ac3ab0bc64a1368c5eedef4cdf04a851e/Dshop/apps/products_catalogue/tests/test_products.py#L13 and https://github.com/pymasterspl/Dshop/blob/dev/Dshop/apps/products_catalogue/tests/test_categories.py#L16
* I don't see a test on category detail page, that it is only listing main products, not variants (you can use the context from above)

Reply:
done